crates/pop-fork/src/cache.rs

```diff
@@ -0,0 +1,564 @@
+		}))
+	}
+
+	/// Cache a storage value.
+	///
+	/// # Arguments
+	/// * `block_hash` - The block hash this storage is from
+	/// * `key` - The storage key
+	/// * `value` - The storage value, or None if the key has no value (empty)
+	pub async fn set(
+		&self,
+		block_hash: H256,
+		key: &[u8],
+		value: Option<&[u8]>,
+	) -> Result<(), CacheError> {
+		// Insert or update the cached storage entry.
+		// Uses INSERT OR REPLACE (SQLite's UPSERT) to handle both new entries
+		// and updates to existing entries with the same (block_hash, key) primary key.
+		// The `is_empty` flag is set based on whether value is None, allowing us
+		// to cache the knowledge that a storage key has no value (vs not being cached).
+		sqlx::query(
+			"INSERT OR REPLACE INTO storage (block_hash, key, value, is_empty) VALUES (?, ?, ?, ?)",
+		)
+		.bind(block_hash.as_bytes())
+		.bind(key)
+		.bind(value)
+		.bind(value.is_none())
+		.execute(&self.pool)
+		.await?;
+
+		Ok(())
+	}
+
+	/// Get multiple cached storage values in a batch.
+	///
+	/// Returns results in the same order as the input keys.
+	pub async fn get_batch(
+		&self,
+		block_hash: H256,
+		keys: &[Vec<u8>],
+	) -> Result<Vec<Option<Option<Vec<u8>>>>, CacheError> {
+		if keys.is_empty() {
+			return Ok(vec![]);
+		}
+
+		// Build a SELECT query with dynamic IN clause for batch retrieval.
+		// This fetches all requested keys in a single round-trip to the database,
+		// which is more efficient than individual queries when fetching many keys.
+		// Example: SELECT ... WHERE block_hash = ? AND key IN (?, ?, ?)
+		let placeholders: Vec<_> = keys.iter().map(|_| "?").collect();
+		let query = format!(
+			"SELECT key, value, is_empty FROM storage WHERE block_hash = ? AND key IN ({})",
+			placeholders.join(", ")
+		);
+
+		let mut query_builder = sqlx::query(&query).bind(block_hash.as_bytes());
+
+		for key in keys {
+			query_builder = query_builder.bind(key.as_slice());
+		}
+
+		let rows = query_builder.fetch_all(&self.pool).await?;
+
+		// Build a map from the results. SQLite doesn't guarantee result order matches
+		// the IN clause order, so we use a HashMap to look up values by key.
+		let mut cache_map: std::collections::HashMap<Vec<u8>, Option<Vec<u8>>> =
+			std::collections::HashMap::new();
+		for row in rows {
+			let key: Vec<u8> = row.get("key");
+			let is_empty: bool = row.get("is_empty");
+			let value = if is_empty { None } else { Some(row.get("value")) };
+			cache_map.insert(key, value);
+		}
+
+		// Return values in the same order as input keys.
+		// Keys not found in cache_map (not in DB) return None (not cached).
+		// Keys found return Some(value) where value is None for empty or Some(bytes) for data.
+		Ok(keys.iter().map(|key| cache_map.get(key).cloned()).collect())
+	}
+
+	/// Cache multiple storage values in a batch.
+	///
+	/// Uses a transaction for efficiency.
+	pub async fn set_batch(
+		&self,
+		block_hash: H256,
+		entries: &[(Vec<u8>, Option<Vec<u8>>)],
```

Comment:
```suggestion
		entries: &[(&[u8], Option<&[u8]>)],
```

Same reasons as in `get_batch`